src/coreclr/vm/unsafeaccessors.cpp

```diff
@@ -75,16 +93,258 @@ namespace
 
         UnsafeAccessorKind Kind;
         MethodDesc* Declaration;
-        MetaSig DeclarationSig;
+        SigPointer DeclarationSig;  // This is the official declaration signature. It may be modified
+                                    // to include the UnsafeAccessorTypeAttribute types.
+        MetaSig DeclarationMetaSig;
+        NewArrayHolder<ParamDetails> TranslatedParams; // Redefined types from UnsafeAccessorTypeAttribute usage.
+                                                       // Return type is at 0 index. Function arguments are 1 to N.
         SigPointer TargetTypeSig;
         TypeHandle TargetType;
         bool IsTargetStatic;
         MethodDesc* TargetMethod;
         FieldDesc* TargetField;
     };
 
+    void AppendTypeToSignature(
+        SigBuilder& sig,
+        TypeHandle th,
+        Instantiation inst)
+    {
+        STANDARD_VM_CONTRACT;
+        _ASSERTE(!th.IsNull());
+
+        //
+        // Building the signature follows details defined in ECMA-335 - II.23.2.12
+        //
+
+        // If we have any generic variables, mark as ELEMENT_TYPE_GENERICINST.
+        BOOL hasGenericVariables = !inst.IsEmpty();
+        if (hasGenericVariables)
+        {
+            sig.AppendElementType(ELEMENT_TYPE_GENERICINST);
+
+            // Embed the generic type definition in the signature.
+            th = TypeHandle{ th.GetMethodTableOfRootTypeParam()->GetTypicalMethodTable() };
+            _ASSERTE(th.IsGenericTypeDefinition());
+        }
+
+        // Append the new type to the signature.
+        sig.AppendElementType(ELEMENT_TYPE_INTERNAL);
+        sig.AppendPointer(th.AsPtr());
+
+        // Append the instantiation types to the signature.
+        if (hasGenericVariables)
+        {
+            _ASSERTE(inst.GetNumArgs() > 0);
+            sig.AppendData(inst.GetNumArgs());
+            for (DWORD i = 0; i < inst.GetNumArgs(); i++)
+            {
+                TypeHandle& instTh = inst[i];
+                if (instTh.IsGenericVariable())
+                {
+                    TypeVarTypeDesc* typeVar = instTh.AsGenericVariable();
+                    sig.AppendElementType(typeVar->GetInternalCorElementType());
+                    sig.AppendData(typeVar->GetIndex());
+                }
+                else if (CorIsPrimitiveType(instTh.GetSignatureCorElementType()))
+                {
+                    sig.AppendElementType(instTh.GetSignatureCorElementType());
+                }
+                else
+                {
+                    sig.AppendElementType(ELEMENT_TYPE_INTERNAL);
```

Comment:
I think this needs to run recursively to handle things like:

`List<List<!0[]>>`